File: ports/dotnet/src/RussianDocs.DocumentProcessing/Pipeline/Timings.cs
using System.Diagnostics;
using RussianDocs.DocumentProcessing.Tensors;

namespace RussianDocs.DocumentProcessing.Pipeline;

/// <summary>
/// Per-stage timings.
///
/// <para>
/// **The stage NAMES are part of the compared contract**, which is not obvious: the tolerance spec
/// says timings are never compared by VALUE, and that is true — but the comparator descends into
/// dictionaries before consulting its ignore list, so the SET OF KEYS is compared exactly.
/// </para>
///
/// <para>
/// The names come from the reference's private method names, leading underscore and all. Renaming
/// them to something idiomatic would be a breaking change, not a tidy-up.
/// </para>
/// </summary>
public sealed class Timings
{
    public const string DocTypeAngle = "_doctype_angle";
    public const string QualityAndBorders = "_quality_and_borders";
    public const string Glare = "_glare";
    public const string Blur = "_blur";
    public const string PrintSpoofing = "_print_spoofing";
    public const string LcdSpoofing = "_lcd_spoofing";
    public const string DocDetector = "_doc_detector";
    public const string Deskew = "_deskew";
    public const string FieldsDetector = "_fields_detector";
    public const string SplitWords = "_split_words";
    public const string Ocr = "_ocr";

    private readonly Dictionary<string, double> _stages = new(StringComparer.Ordinal);
    private readonly HashSet<string> _concurrent = new(StringComparer.Ordinal);

    public void Record(string stage, TimeSpan elapsed) =>
        _stages[stage] = Ops.RoundHalfEven(elapsed.TotalSeconds, 4);

    /// <summary>Times an action and records it under <paramref name="stage"/>.</summary>
    public T Time<T>(string stage, Func<T> action)
    {
        var watch = Stopwatch.StartNew();
        try
        {
            return action();
        }
        finally
        {
            // Recorded in a finally block, so a failing stage still reports how long it took before
            // failing — which is the timing an operator actually wants when something is slow.
            Record(stage, watch.Elapsed);
        }
    }

    public void Time(string stage, Action action) => Time(stage, () =>
    {
        action();
        return 0;
    });

    /// <summary>
    /// Records a concurrent group: its own wall time, plus each member's.
    ///
    /// <para>
    /// The members are marked concurrent so they do NOT contribute to the total — the group's wall
    /// time already covers them, and adding both would report more time than elapsed.
    /// </para>
    /// </summary>
    public void RecordGroup(string name, TimeSpan wall, IReadOnlyDictionary<string, TimeSpan> members)
    {
        Record(name, wall);
        foreach ((string stage, TimeSpan elapsed) in members)
        {
            Record(stage, elapsed);
            _concurrent.Add(stage);
        }
    }

    /// <summary>Every stage plus <c>total</c>, which sums only the non-concurrent ones.</summary>
    public Dictionary<string, double> Report()
    {
        var report = new Dictionary<string, double>(_stages, StringComparer.Ordinal);
        double total = _stages.Where(kv => !_concurrent.Contains(kv.Key)).Sum(kv => kv.Value);
        report["total"] = Ops.RoundHalfEven(total, 4);
        return report;
    }
}

File: ports/go/internal/docproc/pipeline/timings.go
package pipeline

import (
	"time"

	"github.com/protei300/RussianDocsOCR/ports/go/internal/docproc/tensor"
)

// Stage keys for the timings report.
//
// These are WIRE NAMES and part of the compared contract: the checker ignores timing
// VALUES but still compares the timings key SET, so a port that invents its own stage
// names fails. In Python they fall out of `func.__name__` on the pipeline's private
// methods, which is why they are snake_case with a leading underscore — a Go-idiomatic
// renaming here would be a breaking change, not a tidy-up (CONVENTIONS §1).
const (
	StageDocTypeAngle      = "_doctype_angle"
	StageQualityAndBorders = "_quality_and_borders"
	StageGlare             = "_glare"
	StageBlur              = "_blur"
	StagePrintSpoofing     = "_print_spoofing"
	StageLcdSpoofing       = "_lcd_spoofing"
	StageDocDetector       = "_doc_detector"
	StageDeskew            = "_deskew"
	StageFieldsDetector    = "_fields_detector"
	StageSplitWords        = "_split_words"
	StageOcr               = "_ocr"
)

// Timings accumulates per-stage wall times.
//
// Port of PipelineResults._timings plus its `timings` property and add_concurrent_group.
type Timings struct {
	stages map[string]float64
	// concurrent names the stages that ran INSIDE a group. They keep their individual
	// times in the report but are excluded from `total`.
	concurrent map[string]bool
}

func NewTimings() *Timings {
	return &Timings{stages: map[string]float64{}, concurrent: map[string]bool{}}
}

// Record stores one stage's elapsed time, rounded to 4 places like the reference.
func (t *Timings) Record(stage string, elapsed time.Duration) {
	t.stages[stage] = tensor.RoundHalfEven(elapsed.Seconds(), 4)
}

// Time runs fn and records how long it took. The equivalent of _model_call, whose only
// job in the reference is exactly this.
func (t *Timings) Time(stage string, fn func() error) error {
	start := time.Now()
	err := fn()
	t.Record(stage, time.Since(start))
	return err
}

// RecordGroup stores a concurrent group: its own elapsed time under `name`, and each
// member's individual time marked as concurrent.
//
// Why the group's own time counts and the members' do not: summing overlapping members
// would put `total` ABOVE the real processing time and, worse, would keep it flat when
// parallelisation actually saves time — hiding the very thing the group exists for.
func (t *Timings) RecordGroup(name string, wall time.Duration, members map[string]time.Duration) {
	t.Record(name, wall)
	for stage, d := range members {
		t.Record(stage, d)
		t.concurrent[stage] = true
	}
}

// Report returns the stage times plus `total`.
//
// `total` covers TIMED STAGES ONLY — image loading and resizing are not a timed stage in
// the reference either, so wall-clock time is slightly higher. Reproduced rather than
// corrected: the number is on the wire.
func (t *Timings) Report() map[string]float64 {
	out := make(map[string]float64, len(t.stages)+1)
	var total float64
	for stage, v := range t.stages {
		out[stage] = v
		if !t.concurrent[stage] {
			total += v
		}
	}
	out["total"] = tensor.RoundHalfEven(total, 4)
	return out
}
